Multi-selection: allow partial block selection

ellatrix opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description

Fixes #27481.
See also #3629.

These changes allow the user to select multiple blocks without selecting the blocks as a whole. This is done by allowing native selection across block boundaries, detecting the selection start and end in the appropriate rich text instances, and handling actions for this selection.

This PR implements `Enter`, `Backspace`, `Delete` (forward merge) and any input (typing) over the selection.

You can select partially by mouse (drag), shift + click, and also the keyboard (shift + arrow).

It works in all browsers including Firefox.

![cross-block-selection](https://user-images.githubusercontent.com/4710635/154561252-39daccae-6329-45fd-b589-00509c6a0c83.gif)

We use the `merge` function on the block type to handle merges and detect if two blocks are mergeable. If two blocks are not mergeable, we fall back to the current behaviour, which is to select the blocks entirely.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->
- [x] I've updated related schemas if appropriate. <!-- Reference: https://github.com/WordPress/gutenberg/tree/trunk/schemas -->

ellatrix committed:
Initial attempt

ellatrix commented on packages/block-editor/src/components/block-tools/index.js:
```diff
@@ -106,11 +118,33 @@ export default function BlockTools( {
+				mergeBlocks(
+					first( clientIds ),
+					last( clientIds ),
+					clientIds.slice( 1, -1 )
```
Ideally we should just be able to pass `clientIds`.

ellatrix committed:
Fix selection from bottom to top

ellatrix committed:
Ensure native selection spans multiple blocks

ellatrix committed:
Fix isSelected

ellatrix committed:
Fix ctrl+A

mcsf commented:
🎉 

Let us know when you'd like some testing and reviewing done. For now this was really cool to try out, but I can also consistently crash the editor if I delete a span of text across two blocks (using <kbd>Enter</kbd>) and then use the editor's undo action.

ellatrix committed:
Fix e2e tests

ellatrix committed:
Fix text color button

ellatrix committed:
Fix early return

ellatrix committed:
Prevent default behaviour for everything except Backspace and Delete

ellatrix commented:
> For now this was really cool to try out, but I can also consistently crash the editor if I delete a span of text across two blocks (using Enter) and then use the editor's undo action.

That's because `Enter` is not implemented yet. Only `Backspace` and (forward) `Delete` is implemented right now. For now I disabled all native behaviour, which was causing the errors.

ellatrix commented:
Almost all e2e tests should be passing now. What remains to be done is polishing up the code and reusing more of the existing merge function (merge the `mergeBlocks` code paths into a single one).

As part of that, or after that, we can look into adding `Enter` handling. This can be either:

* Merging the blocks first, then split
* Reusing parts of the merge (delete all selected content), then insert a paragraph between the two blocks instead of merging them.

ellatrix committed:
Polish

ellatrix committed:
Fix forward delete

ellatrix committed:
Fix typo

mtias commented on packages/block-editor/src/components/block-tools/index.js:
```diff
@@ -106,13 +94,15 @@ export default function BlockTools( {
-				event.preventDefault();
-				removeBlocks( clientIds );
+
+			if ( clientIds.length < 2 ) {
```
Some of these cases could use a comment to describe the situation we are addressing

mtias commented on packages/rich-text/src/component/use-input-and-selection.js:
```diff
@@ -254,32 +273,23 @@ export function useInputAndSelection( props ) {
 		element.addEventListener( 'compositionstart', onCompositionStart );
 		element.addEventListener( 'compositionend', onCompositionEnd );
 		element.addEventListener( 'focus', onFocus );
-		element.addEventListener( 'blur', onBlur );
```
What were blurs aiming to achieve before?

ellatrix replied:
Removing `selectionchange` event listeners. But now we care about selection outside of rich text as well. More precisely selection that starts in rich text and ends in another. This selection state needs to be communicated to the block editor store so we can perform the keyboard actions on the selection.

mtias commented:
Excellent to see this! I'm glad it's able to build incrementally on the foundation we have. Ensuring `mergeBlocks` is doing its job well as a basis makes sense.

Outside of the mechanics, I think we need to revisit the multiselect outline to only show when you are performing a block level operation (moving multiple blocks, transforming, etc) similar to how we highlight the outline on hover / focus of the blocktype area on single blocks. It might need some design tweaks as well.

One last comment: I think it might be worth (as a follow up) to consider whether it makes sense to allow pressing shift while selecting across blocks to force full block selection instead of partial. Curious what you think about that or if it'd even be necessary given block level tools would still apply block level even with partial selections (inspector controls, etc).

ellatrix commented:
> One last comment: I think it might be worth (as a follow up) to consider whether it makes sense to allow pressing shift while selecting across blocks to force full block selection instead of partial. Curious what you think about that or if it'd even be necessary given block level tools would still apply block level even with partial selections (inspector controls, etc).

Yes I'm definitely going to work on enabling this for keyboard as well. So far focus has been on getting tests to pass :D

ellatrix committed:
Avoid DOM checks

ellatrix committed:
Fix unit test

ellatrix committed:
Allow partial selection with shift + arrow keys

ellatrix committed:
Fix e2e tests

ellatrix committed:
Further fixes

ellatrix committed:
Fix forward delete

ellatrix commented:
> Yes I'm definitely going to work on enabling this for keyboard as well.

Added selection by keyboard.

To do: add `Enter` handling and more e2e tests. Also split the `deleteSelection` code into more parts because it's quite verbose.

ellatrix committed:
Add Enter handling

dmsnell commented on packages/block-editor/src/store/actions.js:
```diff
@@ -658,6 +658,302 @@ export const synchronizeTemplate = () => ( { select, dispatch } ) => {
+	const selectionAnchor = select.getSelectionStart();
+	const selectionFocus = select.getSelectionEnd();
+
+	if ( selectionAnchor.clientId === selectionFocus.clientId ) return;
```
should we not wrap these in `{}`s to avoid making it easy to recreate some infamous bugs due to the missing brackets?

```ts
// oops!
if ( … )
	someNewThing();
	return;
```

ellatrix replied:
Prettier will decrease indentation, so imho it's clear when you add lines. If we don't like this style, we should probably add the relevant rule to the linter.

dmsnell replied:
just raising it because I thought the practice was long abandoned because of its risk. I hadn't seen it in Gutenberg before and it seems riskier still to rely on the auto-formatter to avoid code patterns that invite bugs. carry on 🙇‍♂️

(by the way, it would seem fine to add this to the linter if other people are also leery of single-line `if` blocks)

ellatrix replied:
Right, I'm fine with avoiding it. Maybe we should add it to the linter though, so it's "official".

dmsnell commented on packages/block-editor/src/store/actions.js:
```diff
@@ -658,6 +658,302 @@ export const synchronizeTemplate = () => ( { select, dispatch } ) => {
+	);
+
+	// It's not mergeable if the selection doesn't start and end in the same
+	// block list. Maybe in the future it should be allowed.
```
is this specific to block list?
if that's so, couldn't two blocks from different block lists have the same `rootClientId === ''`?

it's also about nested block then isn't it? if so, would this be clearer to say something like "…start and end in the same parent block"?

ellatrix replied:
No two block lists can have the same root client ID I think.

dmsnell replied:
okay, it seemed like if they were created independently then each would have the empty string as the root id. thanks for double-checking.

dmsnell commented on packages/block-editor/src/store/actions.js:
```diff
@@ -658,6 +658,302 @@ export const synchronizeTemplate = () => ( { select, dispatch } ) => {
+	// is to insert a special character at the position and then recover it.
+	const START_OF_SELECTED_AREA = '\u0086';
+
+	valueA = insert( valueA, '', selectionA.offset, valueA.text.length );
```
what does inserting the empty string do here?

ellatrix replied:
Yeah this code isn't polished, that's why it's still in draft stage. 😄 This should change to `remove`, which is just `inserting` and empty string underneath.

dmsnell commented on packages/block-editor/src/store/actions.js:
```diff
@@ -978,10 +1259,10 @@ export function exitFormattedText() {
- * @param {string} attributeKey The selected block attribute key.
- * @param {number} startOffset  The start offset.
- * @param {number} endOffset    The end offset.
+ * @param {string|Object} clientId     The selected block client ID.
```
should we update the description here, or better, use a different attribute name for `clientId` if we aren't actually passing a `clientId`? I'm guessing we're looking at passing an actual block instead of a `clientId`, but I would be surprised if I had to do this: `block = action.clientId;`

ellatrix replied:
Right, I just haven't stared fixing all the doc blocks yet.

ntsekouras replied:
➕ to update the docs and possibly the param name here. Also can we rename the passed object's property from `start` to `selectionStart` to map the name in state?

dmsnell reviewed (commented):
No telling how far-reaching and positive this change will be. I'm only starting to digest the changes and hope to provide some more-useful feedback, but as a first pass I'm just trying to understand more than review. Thanks for attacking this problem.

youknowriad reviewed (commented):
This is a great start, people will love to have this. I didn't dive into the code but I wanted to try it a bit. 

Things I noticed so far:

 - It seems the selection works well for paragraph blocks only
 - When I tried, list, quote blocks, selecting across two blocks does some weird things: The selection changes when I release the mouse.
 - After using the feature (Enter after cross paragraph selection), I think the second paragraph block end up in a weird state. I'm not able to use "Enter" on that paragraph anymore, sometimes cross-selection doesn't work anymore.
 - In Firefox, multi-selection doesn't seem to work.
 - I have some blocks that can fail (JS errors) when I try merge or undo after a cross-selection removal or "Enter" (I'm not sure yet about the exact step but here's what I see on the console.
 
<img width="1190" alt="Screen Shot 2022-02-26 at 7 59 55 AM" src="https://user-images.githubusercontent.com/272444/155835407-f3fb2849-500f-4df1-b048-b19506c602c0.png">

mtias commented:
@ellatrix sorry :) to clarify my comment regarding shift, I meant combining shift with pointer based selection (so you drag between blocks while holding shift) to force pure multi-selection. Though I don't think it'd be necessary if we always have block selection happening anyways.

ellatrix commented:
@youknowriad 

> It seems the selection works well for paragraph blocks only
> When I tried, list, quote blocks, selecting across two blocks does some weird things: The selection changes when I release the mouse.

If works for me. Did you pull the most recent changes? Both list and quote blocks have merging functions and I checked that this works correctly. For _some_ blocks that don't have merging functions we will select the entire blocks at mouse up, which is what we do for all blocks currently. That is to communicate to the user that the blocks are selected as a whole (not partially).

> After using the feature (Enter after cross paragraph selection), I think the second paragraph block end up in a weird state. I'm not able to use "Enter" on that paragraph anymore, sometimes cross-selection doesn't work anymore.

I'm still able to press Enter in both the newly created empty paragraph (sitting in the middle) and the third paragraph. Not sure if we're at the same commit 😅

> In Firefox, multi-selection doesn't seem to work.

Confirmed. I will investigate this.

> I have some blocks that can fail (JS errors) when I try merge or undo after a cross-selection removal or "Enter" (I'm not sure yet about the exact step but here's what I see on the console.

This also indicates to me that you're maybe at an earlier state of this PR? I used to get error messages, but I no longer get them.

youknowriad commented:
@ellatrix I think the last Firefox fix broke the behavior in Safari and Chrome. Also, in Firefox, I noticed that I'm not able to start a selection from the currently selected block.

ellatrix committed:
Remove lines, fix occasional selection failure

ellatrix committed:
Stabilise new e2e tests

ellatrix committed:
Allow input over selection

dmsnell commented:
In case anyone's interested, I had a hard time assessing the clipboard contents (as usual when dealing with the clipboard) while testing with this. I made a simple [clipboard viewer](https://dmsnell.github.io/ghpages/clipboard.html) to reveal all the MIME types and previews of the items in the paste events.

ellatrix committed:
Remove obsolete comment

youknowriad commented:
This is getting more solid 👏 Here a couple of issues I found now:

 - If I try to merge into a quote block, I see two issues: The cursor is not put in the right position (meaning writing write in the wrong position) and also there's an extra paragraph for me, I expected to merge with the latest paragraph.
 - If I try to merge into a list block (paragraph into list), there's an extra list item as well, the cursor is well placed though

jasmussen commented on packages/block-editor/src/components/block-list/style.scss:
```diff
@@ -27,8 +27,7 @@
 	.block-editor-block-list__block.is-highlighted,
-	.block-editor-block-list__block.is-multi-selected {
-
+	.block-editor-block-list__block.is-highlighted ~ .is-multi-selected {
```
This change makes blocks look partially selected, like so:
![before](https://user-images.githubusercontent.com/1204802/158360813-7ac5db0b-5955-4638-8534-6c2e311c8978.gif)

If we change this rule to the following instead, it appears to thread the needle for me:

`.block-editor-block-list__block.is-multi-selected:not(.rich-text) {`

![after](https://user-images.githubusercontent.com/1204802/158360910-c4664097-f301-49c2-9728-2cea272678da.gif)

What do you think?

mtias replied:
Yes, that seems better. Can we check it still provides the right result for quotes, lists, headings, etc?

ellatrix committed:
Remove scroll to multi-selection

jasmussen committed:
Polish highlight and block select style.

This commit does a few things:

* Changes the highlight style to be 1.5px like the selected styles. This is actually an oversight in trunk, that's just fixed here.
* Adds non-text blocks to the list of blocks that get a hidden selection style. This makes it clearer when you're selecting between two paragraphs, but across a block in between them.

ntsekouras commented on packages/block-editor/src/store/actions.js:
```diff
@@ -658,6 +658,302 @@ export const synchronizeTemplate = () => ( { select, dispatch } ) => {
+
+	dispatch.replaceBlocks(
+		select.getSelectedBlockClientIds(),
+		[ cloneA, createBlock( 'core/paragraph' ), cloneB ],
```
We should use `getDefaultBlockName` here.

ntsekouras commented on packages/block-editor/src/store/actions.js:
```diff
@@ -658,6 +658,302 @@ export const synchronizeTemplate = () => ( { select, dispatch } ) => {
+};
+
+export const splitSelection = () => ( { select, dispatch } ) => {
+	const selectionAnchor = select.getSelectionStart();
```
Do you think it's worth extracting some of this logic as is reused in a couple of places? The logic should be the same across split, delete, no?

ellatrix replied:
Yeah. It's probably good to split into multiple parts anyway because it's quite verbose

ntsekouras commented on packages/block-editor/src/store/actions.js:
```diff
@@ -658,6 +658,302 @@ export const synchronizeTemplate = () => ( { select, dispatch } ) => {
+	const focusIndex = blockOrder.indexOf( selectionFocus.clientId );
+
+	// Reassign selection start and end based on order.
+	let selectionStart, selectionEnd;
```
I guess we don't need these vars and we could instead use `selectionA` and `selectionB`.

ntsekouras commented on packages/block-editor/src/store/reducer.js:
```diff
@@ -1298,6 +1305,14 @@ export function selection( state = {}, action ) {
 		case 'MULTI_SELECT':
 			const { start, end } = action;
+
+			if (
```
Why do we need this?

ellatrix replied:
Because it might overwrite more specific selection with an offset. Some tests were failing without this, but forgot which ones :)

ntsekouras reviewed (commented):
Impressive work here Ella! I'll continue testing and I've left some minor comments now.

ellatrix committed:
Fix for Firefox

ellatrix committed:
Fix keyboard nav for Firefox

youknowriad commented:
Another bug I noticed, if I have a 

paragraph
image
paragraph


And I try to select from bottom to top, sometimes, it doesn't work (inconsistent).

youknowriad commented:
I confirm that in Firefox it's way better. That said, I noticed a small bug. If the block where I release the mouse is "quote" or "list", the multi selection is lost.

ellatrix commented:
> Another bug I noticed, if I have a
> 
> paragraph image paragraph
> 
> And I try to select from bottom to top, sometimes, it doesn't work (inconsistent).

Do you mean for Firefox or all browsers? Seems to work well in Chrome. For Firefox I can select upwards, but on mouse up the selection will fill the blocks.

ellatrix commented:
<img width="337" alt="Screenshot 2022-03-16 at 15 04 40" src="https://user-images.githubusercontent.com/4710635/158596203-fa336a33-3946-4159-b693-b2e00513732b.png">

I can select into a list in Firefox though.

ellatrix committed:
Use getDefaultBlockName

ellatrix committed:
insert '' => remove

Mamaduka commented:
PR tests well in Safari.

> If the block where I release the mouse is "quote" or "list", the multi selection is lost.

I see the same behavior in Safari with the "Pullquote" block.

https://user-images.githubusercontent.com/240569/158603284-f9c02624-8fc7-4593-a71b-f802c8d85974.mp4

youknowriad commented:
> Do you mean for Firefox or all browsers? Seems to work well in Chrome. For Firefox I can select upwards, but on mouse up the selection will fill the blocks.

All browsers, here's a screenshot in safari

https://user-images.githubusercontent.com/272444/158607285-df22309b-9039-4864-a2fc-8756e046f80f.mov

ellatrix commented:
> * If I try to merge into a quote block, I see two issues: The cursor is not put in the right position (meaning writing write in the wrong position) and also there's an extra paragraph for me, I expected to merge with the latest paragraph.
> * If I try to merge into a list block (paragraph into list), there's an extra list item as well, the cursor is well placed though

So this is because of how the merge function works currently. When merging with a multiline value like quote or list, the merge function joins the values as separate lines. Imo, this is something we have to fine tune with the quote and list block rewrite since they'll no longer be multiline values. With the rewrite there will be two scenarios:

* When the caret is at the beginning of the second block, you merge into the preceding **list** block.
* When you merge a partial selection, you merge into the **list item** block because that's where the selection starts.

youknowriad commented:
> multi-selection lost when you select into a list in firefox


https://user-images.githubusercontent.com/272444/158609040-3ad8bf27-c260-4627-93e5-8d5278b021ba.mov

ellatrix commented:
Ah, you mean an image **placeholder**.

It seems to sometimes works and sometimes not. Not sure what I can do about it right now. I think the idea to use shadow DOM for placeholders could help here to avoid trapping selection. Not sure what the root cause is though.

![selecting-across-image-placeholder](https://user-images.githubusercontent.com/4710635/158611645-a2eb2dbd-7fc1-4f7a-8e92-08c96d4d22f3.gif)

ellatrix committed:
Simplify selection filling

ellatrix commented:
> > multi-selection lost when you select into a list in firefox

@youknowriad Should be fixed with 24f3d1ddc4c95b6ae6ae6153c488282d44a3edc3.

youknowriad commented:
> @youknowriad Should be fixed with https://github.com/WordPress/gutenberg/commit/24f3d1ddc4c95b6ae6ae6153c488282d44a3edc3

What happens for me now is that we always fallback to normal multi-selection (all blocks) in Firefox. That's fine by me though.

ellatrix committed:
Fix Firefox focus issue

ellatrix committed:
Fix e2e tests

ellatrix committed:
Fix selection after merge

ellatrix committed:
Move selection tracking to writing flow

ellatrix committed:
Fix import

ellatrix committed:
Fix e2e test

ellatrix committed:
Separate click to select hook

ellatrix committed:
Fix keyboard selection with non rich text

ellatrix committed:
Restrict filling in selection if block already has selection

ellatrix committed:
Fix paste e2e test

ellatrix committed:
Fix race condition in e2e tests

ellatrix committed:
Allow partial selection with shift+click

youknowriad commented:
I've been testing this and it feels super good now. Here are the few things I noticed:

 - Copy pasting still copies the whole blocks. I think it's fine to ship the V1 of this PR like that.
 - backspace/writing into the quote blocks and list blocks still results in an extra line. (I think that's also fine for V1)
 - backspacing/writing a multi selection into the quote block causing a focus jump to the beginning of the block. I think we should fix that before shipping this as the behavior feels completely broken there.

youknowriad commented:
This behavior felt a bit weird and I think folks might consider this a bug (content loss)

https://user-images.githubusercontent.com/272444/159229512-490840d6-117e-4a5a-968b-3460c70ac4ef.mov

Basically when partially selecting a container block in addition to some blocks outside the container, the whole container is removed.

I think if we want to keep that behavior, we should auto-expand the selection on mouse release / keyboard selection.

jasmussen commented:
> Basically when partially selecting a container block in addition to some blocks outside the container, the whole container is removed.

I'd count "Group" as a non-rich-text block like others, and treat it the same wiy with the blue focus outline then appearing. Same as if you select from a paragraph into an image.

ellatrix commented:
> I think if we want to keep that behavior, we should auto-expand the selection on mouse release / keyboard selection.

@jasmussen @youknowriad We need to find a way to show full selection without native selection (and maybe hide native selection). We can't really expand native selection as some selection state will be lost (selection direction and initial position). But let me tinker with some alternatives like expanding only to one side to keep the direction and initial position.

jasmussen commented:
This worked for the image block with the last commit I pushed a while back, that simply paints shows the blue focus and hides the native selection.

ellatrix committed:
wip

ellatrix committed:
Fix selection issue after quote merge

ellatrix committed:
Fix e2e tests

youknowriad commented:
I've been testing this again and both the blockers I raised have an acceptable solution. I noticed some undo bugs. Sometimes undo doesn't work after removing a multi-selection, but when the multi selection involves a container block, the restored block crashes (not sure it's strictly related to this PR)


https://user-images.githubusercontent.com/272444/159660039-cdc0df52-2bd1-4f8b-89f4-fa5883d35731.mov

<img width="1429" alt="Screen Shot 2022-03-23 at 9 51 34 AM" src="https://user-images.githubusercontent.com/272444/159660147-33a1eb99-6f53-489a-a9cd-4ab15450d647.png">

youknowriad commented on docs/reference-guides/data/data-core-block-editor.md:
```diff
@@ -1468,6 +1489,10 @@ _Returns_
+### splitSelection
+
+Split the current selection.
+
```
How confident are we about the new selector/actions? Mainly asking if we need time for them to be experimental or is it fine to be stable from the get go.

ellatrix replied:
Yeah, probably good to make unstable for now.

youknowriad commented on packages/block-editor/src/components/block-list/use-block-props/use-focus-first-element.js:
```diff
@@ -124,7 +123,11 @@ export function useFocusFirstElement( clientId ) {
 
-		setContentEditableWrapper( ref.current, false );
+		// This is needed because this hook is called before the writing flow
+		// hook has the chance to turn off content editable.
```
Can we clarify in the comment what happens if we remove this (reproduction steps) or maybe we have some e2e test covering this?

youknowriad commented on packages/block-editor/src/components/rich-text/use-focus.js:
```diff
@@ -0,0 +1,39 @@
+ */
+import { store as blockEditorStore } from '../../store';
+
+export function useFocus() {
```
Can we add some documentation, it's hard to know what this hook is doing just by its name.

youknowriad commented on packages/block-editor/src/store/actions.js:
```diff
@@ -658,6 +658,305 @@ export const synchronizeTemplate = () => ( { select, dispatch } ) => {
+ *
+ * @param {boolean} isForward
+ */
+export const deleteSelection = ( isForward ) => ( {
```
I wonder if these two new actions (deleteSelection, splitSelection) deserve their own unit tests. I'm not sure we can test everything here in e2e tests. WDYT

youknowriad reviewed (commented):
This is looking good to me and ready close to land IMO. Now, I'd like to make sure we leave our code in a good situation to address any future needs:

 - Document everything properly (hooks, hacks...).
 - Do we need an architecture doc explaining how multi-selection works in general.
 - Test everything properly to avoid future regressions.
 - Create some follow-up issues for the identified issues: List and blocks adding extra lines for instance, undo-redo bugs.

ellatrix commented:
@youknowriad Yes, I'll add more documentation and a doc explaining things better

ellatrix committed:
wip

ellatrix committed:
Expand selection on Delete if content can't be merged

ellatrix committed:
Refine

ellatrix committed:
Polish

youknowriad reviewed (approved):
I don't know what kind of magic show is this, but I can say that it's a very cool one.

jasmussen commented:
🔥

ellatrix committed:
Revert firefox compat move

ellatrix commented:
Thanks for the reviews! Just fixing the tests and cleaning up a bit, then I will merge.

ellatrix committed:
Revert unnessary changes for this PR

ellatrix committed:
Remove unnessary cloning

ellatrix committed:
Fix typos

mcsf commented on packages/e2e-tests/specs/editor/various/copy-cut-paste-whole-blocks.test.js:
```diff
@@ -100,6 +100,7 @@ describe( 'Copy/cut/paste of whole blocks', () => {
 		await page.click( '.editor-block-list-item-paragraph' );
 		await page.keyboard.type( 'P' );
 		await page.keyboard.press( 'ArrowLeft' );
+		await page.evaluate( () => new Promise( requestAnimationFrame ) );
```
Is this is a Puppeteer smell, or a "us" smell? 😄

ellatrix replied:
This is us. We are doing some delays with `requestAnimationFrame`, so I'll comment

ZebulanStanphill replied:
I wonder if this would stop being an issue after the switch to Playwright?

ellatrix replied:
Not sure. I don't mind adding the delay in tests if we are adding the delay.

mcsf commented on packages/rich-text/src/component/index.js:
```diff
@@ -214,6 +214,19 @@ export function useRichText( {
 		}
 
+		if ( ref.current.ownerDocument.activeElement !== ref.current ) {
+			// Can't focus if there is a parent content editable.
```
Any chance that there would be more than one such ancestor? Or that we should guard against corrupt DOM states?

ellatrix replied:
I will comment more here. We don't have any deeper nesting no :) This fix is actually needed because writing flow doesn't turn off `contenteditable` quick enough. Hooks in child components execute first, and the writing flow hooks last. I wish we had a better fix for this.

ellatrix commented on packages/block-editor/src/components/writing-flow/use-keyboard-selection-observer.js:
```diff
@@ -0,0 +1,142 @@
+					// `selectionchange` listener was added first so it will be
+					// called before the rich text one.
+					rafId = defaultView.requestAnimationFrame( () => {
+						multiSelect( startPath[ depth ], endPath[ depth ] );
```
@mcsf We're using `requestAnimationFrame`. Not sure how to avoid it. The listener callbacks are called in order of which is attached first. The writing flow one is attached first, but we need the rich text one to be executed first.

youknowriad replied:
Note that we can also try `queueMicrotask` which basically will delay the execution after the current event (might not be enough but if it is, it's a bit better than `requestAnimationFrame` as it doesn't require e2e test hack)

ellatrix replied:
`queueMicrotask` wouldn't work here. Here's a demonstration: https://codepen.io/iseulde/pen/JjMWNLa

ellatrix replied:
I removed the RAF by resetting listeners on focus. That way these listeners are always added behind the rich text ones.

ellatrix committed:
Move input handling to writing flow

ellatrix committed:
Fix e2e test

ellatrix committed:
Add more comments

ellatrix committed:
Try removing duplicate code

ellatrix committed:
Make e2e tests pass again

ellatrix committed:
Remove raf hack

ellatrix committed:
Fix block first focus

ellatrix committed:
Fix comments in e2e tests about raf

ellatrix committed:
Fix shift+click selection and add e2e test

ellatrix committed:
Remove unnecessary code

ellatrix committed:
Check to see which tests are failing without this

ellatrix committed:
Check if this is still needed

youknowriad commented:
This is still working good but I'm noticing a couple more issues:

 - I can't select a text inside a quote
 - shift + selection behaves weird (testing on Safari): I shift select multiple blocks starting from a paragraph. Sometimes if I try to delete, selection jumps happen but nothing is deleted, sometimes it just stop multi-selecting... It depends I guess on what blocks I'm selecting (sometimes I include buttons, quote in the selections)

ellatrix commented:
@youknowriad I tested both of your points in different browsers and it works fine. Can select text in Safari, Firefox and Chrome. Shift+Arrow seems to work well in Safari.

![keyboard-selection-safari](https://user-images.githubusercontent.com/4710635/160354674-0ae78627-fe54-4b9c-a152-338480de83f7.gif)

youknowriad commented:
I just tested again and I also can't reproduce. I don't know if I was on the wrong commit or if it something that starts happening after some interactions.

Anyway, found another smaller thing where sometimes I have to hit "backspace" twice to delete a simple selection (inside a quote).

https://user-images.githubusercontent.com/272444/160356919-d8e3e8d6-83af-4741-817b-0d16c75b922e.mov

ellatrix commented:
> Anyway, found another smaller thing where sometimes I have to hit "backspace" twice to delete a simple selection (inside a quote).

Yes, seeing that also. Can't delete selection in list as well. So many behaviours we don't have tests for 🙈

ellatrix committed:
Fix shift+click selection inside block

ellatrix commented:
> Anyway, found another smaller thing where sometimes I have to hit "backspace" twice to delete a simple selection (inside a quote).

I fixed it, just waiting for the tests to run on the last commit to push.

ellatrix committed:
Fix selection inside list and quote blocks

mtias commented:
Outstanding work here, thanks for persevering @ellatrix :)
